src/http/one/ResponseParser.cc

```diff
@@ -100,6 +82,30 @@ Http::One::ResponseParser::parseResponseStatusAndReason(Tokenizer &tok, const Ch
     return -1;
 }
 
+void
+Http::One::ResponseParser::ParseResponseStatus(Tokenizer &tok, StatusCode &code)
+{
+    int64_t statusValue;
+    if (tok.int64(statusValue, 10, false, 3) && tok.skipOne(Parser::DelimiterCharacters())) {
+        debugs(74, 6, "raw status-code=" << statusValue);
+        code = static_cast<StatusCode>(statusValue); // may be invalid
+
+        // RFC 7230 Section 3.1.2 says status-code is exactly three DIGITs
+        if (code <= 99)
+            throw TextException(ToSBuf("status-code too short: ", code), Here());
+
+        // Codes with a non-standard first digit (a.k.a. response class) are
+        // considered semantically invalid per the following HTTP WG discussion:
+        // https://lists.w3.org/Archives/Public/ietf-http-wg/2010AprJun/0354.html
+        if (code >= 600)
+            throw TextException(ToSBuf("status-code from an invalid response class: ", code), Here());
```

Comment:
I hope this status-code semantics enforcement will not trigger a lot of compatibility problems. I think we _should_ be conservative here, at least as long as Squid itself is using zero/6xx status-codes internally and lacks a proper API for response class determination. I anticipate this to be one of the enforcement changes that will need a (configurable) bypass.

Reply:
This enforcement only being on received traffic, not on internal message values should be fine (and correct).

Old Squid used to leak those internal 6xx sometimes. IIRC it was fixed at some point early in 3.x. We should not have many (if any) installations still doing that.

Also some old roundcube versions used to transmit 6xx status in its XHR login traffic. I fixed that some while back too and there have been several major CVE issues triggering upgrades since. So hopefully not many of those left in existence either.

There was some software delivering " 0 " a year or two back. That may be an issue if it is still around. I cannot recall which it was though to check. If we want to be pedantic "000" is three digits (ie valid syntax per section 3.1.2) it is only invalid at semantic layer due to not having a defined meaning.

Reply:
> This enforcement only being on received traffic, not on internal message values should be fine (and correct).
 
My compatibility concern is about _received_ traffic. I mentioned internally used zero/6xx code because that use helps _justify_ the (conservative) changes in this PR. In other words, yes, we should be doing this, but we may have to provide workarounds in the future.

Bugs notwithstanding, internally-created response objects should not have invalid status codes -- invalid status codes should be used internally _outside_ of any response object. Thus, there should be no code that packs internal responses with invalid status codes and, hence, this parsing code should not see any internal responses with invalid status codes.
 
> There was some software delivering " 0 " a year or two back.

FWIW, July 2020 bug [report|https://bugs.squid-cache.org/show_bug.cgi?id=5057#c2] that triggered this PR was allegedly caused by software sending responses with zero status-codes.